enhancements/installer/telemeter-metrics.md

```diff
@@ -0,0 +1,280 @@
+
+### Goals
+
+1. Instrument the installer to collect metrics
+2. Allow the installer to push those metrics to an aggregation pushgateway which is scraped by the Telemetry Prometheus instance
+
+### Non-Goals
+
+1. The metrics collected will be aggregate metrics so metrics will not be relevant to individual installer invocations and will not display detailed data such as log or error messages
+
+## Proposal
+
+### User Stories
+
+
+#### Cluster Admin/Installer User Stories
+As an admin about to install an OpenShift cluster I want to be able to disable metrics collection so that I can opt out.
+
+
+As an admin installing an OpenShift cluster I want my installation result to be unaffected by the success or failure of collecting metrics. 
+
+#### Telemetry Analyst
+Some examples:
+
+
+As a telemetry analyst I would like to create queries which help determine how long the installation process takes in total and broken down by different stages of installation so that I can better understand user experience and expose any abnormalities.
+
+
+As a telemetry analyst I would like to create queries which show aggregate counts of failure by stage of installation so that we can better understand where users are experiencing problems.
+
+
+As a telemetry analyst I would like to see how often users modify ignition configs or manifests so that I can better understand how users are interacting with the installer. [(see open question #4)](open-questions)
+
+
+As a telemetry analyst I would like to see how often the installer is being run on certain operating systems so I can better understand our user environments and needs.
+
+### Implementation Details/Notes/Constraints
+The installer does not fit a typical Prometheus use case in that:
+* metrics will be pushed rather than scraped
+* meaning will come from metric labels as much as values
+* the aggregation push gateway will be used to form the metrics into a useful time series
+
+Initially, three proposed metrics could provide value:
+
+#### cluster_installation_invocation
+The primary metric `cluster_installation_invocation` would be pushed when users invoke either the `create`, `destroy`, `gather`, or `wait-for` command. The values for the labels would be collected throughout the execution of the command and converted to a metric before being pushed.  The `duration` of the command will be bounded in the installer to preset buckets of a certain step-size, 2.5 in the example shown here:
+```
+# HELP cluster_installation_invocation represents a single command run by the OpenShift installer and related outcomes.
+# TYPE cluster_installation_invocation counter
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="ClusterComplete",return="0",version="4.2",duration="25"} 1
```

Comment:
The idea is to bound cardinality by binning the values on the installer client side, similar to a histrogram but not inclusive. e.g. we could have values that represent that installation took: less than 15 minutes, 20 minutes, 25 minutes, 30 minutes, 35 minutes, 40 minutes, more than 45 minutes.

So that example would have 7 potential values. We would discuss what values would be best.

If that doesn't resolve the cardinality issues let me know and we can discuss alternative approaches.

Reply:
Just having a bound cardinality sounds good to me, but I think that depends on how much you can send via telemetry, there are certain budgets per component. 


It's also depends on what is actually useful to you when analyzing this data. If 5 minute interval values are important then this makes sense. The same goes for all the rest of the labels.

Reply:
Why not create exactly a histogram format? This looks like something custom: `duration` vs official `le` label.

Reply:
> Why not create exactly a histogram format? This looks like something custom: duration vs official le label.
+1 see: https://github.com/openshift/enhancements/pull/205#discussion_r381388713